### DISCUSSION TITLE
how to return error with data from Aff

### DISCUSSION THREAD

**Discussion by harrhp** [louthy/language-ext#905] [General]
Right now i usually return `EitherAsync<AppError, A>` from my services,
```
interface AppError {}

record AppError1 : AppError;

record AppError2(string SomeValue) : AppError;
```

so in controller i can match on either and switch on error type to access properties in `AppError2` case. I've read wiki about Aff, but there is no example of returning error with data. The only way i see how to achieve this with Error is to convert my AppErrors to Exceptions. Is this ok or there is something better? Maybe there should be `@catch` overload
```
public static CatchValue<A> @catch<A>(Func<Exception, Option<Fin<A>>> Fail);
```
if possible to be able to handle different exception types differently and require downcast only in one place?

**Comment by louthy** [comment]
It is an intentional limitation of the `Aff`/`Eff` implementation, because it would mean another 4 monads (to support an `E` error type).  And whilst I'm not against that idea, the amount of typing to get to this point has been enormous, and I'm not sure I've got the energy to do another four monads on-top of the core four monads (at least, not any time soon).

So, yes, the only way to leverage the `Error` type as it stands right now is to make `AppError` into an:
```c#
    abstract class AppError : Exception
```
It does unfortunately mean dropping the use of records, which granted isn't great.
